File: algorithms/named/union_find.py
"""
Union Find (or Disjoint Set)

Page 251-253 of Skiena's *The Algorithm Design Manual*.
"""
import unittest
from typing import List, Tuple

NodeId = int
ParentsList = List[int]


class UnionFind:
    def __init__(self, set_size: int) -> None:
        self.set_size = set_size
        self.parents: ParentsList = [
            i for i in range(0, self.set_size + 1)
        ]  # Handle 1-based indexing.
        self.tree_sizes = [i for i in range(0, self.set_size + 1)]

    def root_of(self, i: int) -> int:
        if self.parents[i] == i:
            return i
        return self.root_of(self.parents[i])

    def union(self, left: int, right: int) -> None:
        root_left = self.root_of(left)
        root_right = self.root_of(right)

        if root_left == root_right:
            return  # Already connected / unioned.

        # To minimize tree height, we want to connect the shorted subtree
        # to the longer subtree. This will ensure overall tree height increases
        # by at most one, maybe zero.
        if self.tree_sizes[root_left] >= self.tree_sizes[root_right]:
            self.tree_sizes[root_left] = (
                self.tree_sizes[root_left] + self.tree_sizes[root_right]
            )
            self.parents[root_right] = root_left
        else:
            self.tree_sizes[root_right] = (
                self.tree_sizes[root_right] + self.tree_sizes[root_left]
            )
            self.parents[root_left] = root_right

    def is_connected(self, left: int, right: int) -> bool:
        return self.root_of(left) == self.root_of(right)

    def process_edges(self, edges: List[Tuple[int, int]]) -> None:
        for edge in edges:
            self.union(left=edge[0], right=edge[1])


class TestUnionFind(unittest.TestCase):
    def test_correctness(self):
        uf = UnionFind(set_size=8)
        edges = [(1, 0), (0, 2), (5, 3), (3, 4), (6, 7)]
        uf.process_edges(edges)
        self.assertTrue(uf.is_connected(1, 1))
        self.assertTrue(uf.is_connected(1, 0))
        self.assertTrue(uf.is_connected(6, 7))
        self.assertTrue(uf.is_connected(5, 4))  # Connected by way of 3


if __name__ == "__main__":
    unittest.main()

File: algorithms/named/UnionFind.java
package algorithms.named;

public class UnionFind {
    private static int defaultMaxSize = 100;
    private int[] parents;
    private int[] sizes;
    private int n;  // Number of elems in set.

    public UnionFind() {
        this(defaultMaxSize);
    }

    public UnionFind(int setSize) {
        this.parents = new int[setSize+1];
        this.sizes = new int[setSize+1];
        this.n = setSize;

        for (int i = 1; i <= this.n; i++) {
            this.parents[i] = i;
            this.sizes[i] = 1;
        }
    }

    public int rootOf(int i) {
        if (i > this.n) {
            throw new IllegalArgumentException();
        }

        if (this.parents[i] == i) {
            return i;
        }
        return rootOf(this.parents[i]);
    }

    public void union(int left, int right) {
        int rootRight;
        int rootLeft;

        rootLeft = rootOf(left);
        rootRight = rootOf(right);

        if (rootLeft == rootRight) {
            return; // already connected!
        }

        // Connect the shorter subtree to the longer,
        // to minimize tree height.
        if (this.sizes[left] >= this.sizes[right]) {
            // Make right a subtree of left.
            this.sizes[left] = this.sizes[right] + this.sizes[left];
            this.parents[rootRight] = rootLeft;
        } else {
            // Make left a subtree of right.
            this.sizes[right] = this.sizes[right] + this.sizes[left];
            this.parents[rootLeft] = rootRight;
        }
    }

    public boolean isConnected(int left, int right) {
        int rootLeft = rootOf(left);
        int rootRight = rootOf(right);
        return (rootLeft == rootRight);
    }

    public void processEdges(int[][] edges) {
        int[] edge;
        for (int edgeIdx = 0; edgeIdx < edges.length; edgeIdx++) {
            edge = edges[edgeIdx];
            union(edge[0], edge[1]);
        }
    }
}
